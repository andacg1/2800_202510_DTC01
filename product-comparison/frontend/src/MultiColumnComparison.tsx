import React, { useContext, useEffect, useState } from "react";
import type { ActionMeta, MultiValue, StylesConfig } from "react-select";
import Select from "react-select";
import { isAvailable, type Product } from "./product.ts";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCheck, faX } from "@fortawesome/free-solid-svg-icons";
import { LocationContext } from "./LocationContext.ts";
import makeAnimated from "react-select/animated";
import { RecommendationContext } from "./RecommendationQuery/RecommendationContext.ts";
import { getShortId } from "./utils.ts";

/**
 * Props for the MultiColumnComparison component
 */
type MultiColumnComparisonProps = {
  className?: string;
  children?: React.ReactNode;
  products: Product[];
};

/**
 * Type definition for product options used in the multi-select dropdown
 */
export type ProductOption = {
  readonly value: string;
  readonly label: string;
  readonly product: Product;
};

const animatedComponents = makeAnimated();

/**
 * Custom styles configuration for the React-Select component
 */
const selectStyles: StylesConfig<ProductOption, true> = {
  control: (styles) => ({
    ...styles,
    backgroundColor: "white",
    animation: "none",
    opacity: 1,
    transform: "translateY(0px)",
  }),
  menu: (styles) => ({
    ...styles,
    zIndex: "100",
    backgroundColor: "white",
    animation: "none",
    opacity: 1,
    transform: "translateY(0px)",
  }),
  input: (styles) => ({
    ...styles,
    ":focus-visible": {
      ...styles[":focus-visible"],
      boxShadow: "none",
    },
  }),
};

/**
 * A component that renders a multi-column product comparison interface.
 * Features a multi-select dropdown for product selection and displays a detailed
 * comparison table of the selected products' specifications.
 *
 * @param {Object} props - Component props
 * @param {string} [props.className] - Additional CSS classes to apply to the component
 * @param {React.ReactNode} [props.children] - Child elements
 * @param {Product[]} props.products - Array of products available for comparison
 * @returns {JSX.Element} The rendered multi-column comparison component
 */
const MultiColumnComparison = ({
  className,
  children,
  products,
}: MultiColumnComparisonProps) => {
  const pathName = window.location.pathname;
  const productOptions = products.map((product) => ({
    value: String(product.id),
    label: product.title,
    product: product,
    isFixed: window.location.pathname.includes(product.handle),
  }));

  const [selectedOptions, setSelectedOptions] = useState<
    MultiValue<ProductOption>
  >(
    productOptions.filter((option) =>
      window.location.pathname.includes(option.product.handle),
    ),
  );
  const { location: userLocation } = useContext(LocationContext);
  const { recommendation, setRecommendation } = useContext(
    RecommendationContext,
  );

  /**
   * Gets the currently viewed product based on the URL path
   * @returns {ProductOption | undefined} The current product option or undefined if not found
   */
  const getCurrentProduct = () =>
    productOptions.find((option) => pathName.includes(option.product.handle));

  useEffect(() => {
    console.log({ selectedOptions });
    const currentProductOption = productOptions.find((option) =>
      pathName.includes(option.product.handle),
    );
    console.log({ pathName, productOptions, currentProductOption });
  }, [pathName, productOptions, selectedOptions]);

  useEffect(() => {
    if (!recommendation?.recommendedProductId) {
      console.log("No product ID", recommendation);
      return;
    }
    for (const option of productOptions) {
      console.log(
        getShortId(option.product.id),
        getShortId(recommendation.recommendedProductId),
      );
    }
    setSelectedOptions(
      productOptions.filter(
        (option) =>
          getShortId(recommendation.recommendedProductId) ===
          getShortId(option.product.id),
      ),
    );
  }, [recommendation]);

  /**
   * Gets all unique specification keys across the selected products
   * @returns {string[]} Array of unique specification keys
   */
  const getAllSpecKeys = () => {
    const selectedProductData = products.filter((p) =>
      selectedOptions.map((product) => product.value).includes(String(p.id)),
    );
    const allKeys = new Set<string>();

    selectedProductData.forEach((product) => {
      Object.keys(product.specs).forEach((key) => {
        allKeys.add(key);
      });
    });

    return Array.from(allKeys);
  };

  /**
   * Handles changes in product selection from the multi-select dropdown
   * Tracks the comparison event and updates the selected products state
   *
   * @param {MultiValue<ProductOption>} newValue - The newly selected product options
   * @param {ActionMeta<ProductOption>} actionMeta - Metadata about the selection change action
   */
  const handleProductChange = async (
    newValue: MultiValue<ProductOption>,
    actionMeta: ActionMeta<ProductOption>,
  ) => {
    setSelectedOptions(newValue);
    // TODO: Track the comparison
    const currentProductOption: Product = window?.currentProduct;
    if (!currentProductOption) {
      return console.error('"currentProductOption" is not defined');
    }
    const sessionId = document.cookie
      .split("; ")
      .find((row) => row.startsWith("_shopify_s="))
      ?.split("=")[1];
    console.log({ sessionId });
    const collectionId = window?.collection;

    const resp = await fetch(
      `${import.meta.env.VITE_APP_BACKEND_URL}/api/product/comparison/track`,
      {
        method: "POST",
        body: JSON.stringify({
          collectionId,
          originalProductId: currentProductOption?.id,
          comparedProducts: newValue.map((option) => option.product.id),
          sessionId,
        }),
        mode: "cors",
      },
    );
  };

  return (
    <div className={className}>
      <Select
        isMulti
        name="products"
        options={productOptions}
        components={animatedComponents}
        className="basic-multi-select"
        classNamePrefix="select-internal"
        /* menuIsOpen={true} */
        onChange={handleProductChange}
        value={selectedOptions}
        styles={selectStyles}
      />
      {Array.isArray(selectedOptions) &&
      selectedOptions?.length === 0 ? null : (
        <div className="overflow-x-auto">
          {selectedOptions.length > 0 && (
            <div className="comparison-table">
              <h3>Comparison</h3>
              <table className="table">
                <thead>
                  <tr>
                    <th>Specification</th>
                    {selectedOptions.map((selectedProduct) => {
                      const product = selectedProduct.product;
                      return (
                        <th key={product.id} className="text-center!">
                          {product?.title}
                        </th>
                      );
                    })}
                  </tr>
                </thead>
                <tbody>
                  {getAllSpecKeys().map((specKey) => (
                    <tr key={specKey}>
                      {specKey === "available_regions" && userLocation ? (
                        <td>{`Available in ${userLocation?.country_name || "N/A"}?`}</td>
                      ) : (
                        <td>{specKey.replace("_", " ")}</td>
                      )}
                      {selectedOptions.map((selectedProduct) => {
                        const product = selectedProduct.product;
                        const specValue = product?.specs[specKey];
                        const isAvailableField =
                          specKey === "available_regions";
                        if (
                          isAvailableField &&
                          userLocation &&
                          Array.isArray(specValue)
                        ) {
                          return (
                            <td
                              key={`${product.id}-${specKey}`}
                              className="text-center"
                            >
                              <div className="flex flex-col items-center justify-center">
                                {isAvailable(userLocation, specValue) ? (
                                  <FontAwesomeIcon
                                    icon={faCheck}
                                    color={"rgba(31,255,0,0.5)"}
                                    size={"2x"}
                                    fixedWidth
                                  />
                                ) : (
                                  <FontAwesomeIcon
                                    icon={faX}
                                    color={"#b02525"}
                                    size={"2x"}
                                    fixedWidth
                                  />
                                )}
                                <div>{specValue.join(", ")}</div>
                              </div>
                            </td>
                          );
                        }
                        return (
                          <td
                            key={`${product.id}-${specKey}`}
                            className="text-center"
                          >
                            <div className="flex flex-col items-center justify-center">
                              {Array.isArray(specValue)
                                ? specValue.join(", ")
                                : specValue?.toString() || "N/A"}
                            </div>
                          </td>
                        );
                      })}
                    </tr>
                  ))}
                </tbody>
              </table>
            </div>
          )}
        </div>
      )}
    </div>
  );
};

export default MultiColumnComparison;
