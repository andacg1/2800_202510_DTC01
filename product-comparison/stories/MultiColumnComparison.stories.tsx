import React from "react";
import type { Meta, StoryObj } from "@storybook/react";
import MultiColumnComparison from "../frontend/src/MultiColumnComparison";

// Define a minimal type that satisfies the component's requirements
interface ProductLike {
  id: number;
  title: string;
  handle: string;
  specs: Record<string, any>;
  [key: string]: any; // To allow for additional properties
}

// Sample product data for the story
const sampleProducts: ProductLike[] = [
  {
    id: 1,
    title: "Product A",
    handle: "product-a",
    description: "Description for Product A",
    published_at: "2023-01-01",
    created_at: "2023-01-01",
    type: "Type A",
    vendor: "Vendor A",
    tags: ["tag1", "tag2"],
    variants: [],
    images: [],
    options: [],
    shop: "shop.myshopify.com",
    status: "active",
    price: 100,
    price_min: 100,
    price_max: 100,
    available: true,
    compare_at_price: 120,
    inventory_quantity: 10,
    inventory_management: "shopify",
    inventory_policy: "deny",
    fulfillment_service: "manual",
    requires_shipping: true,
    taxable: true,
    weight: 1,
    weight_unit: "kg",
    specs: {
      color: "Red",
      size: "Medium",
      weight: "2kg",
      available_regions: ["US", "CA", "EU"],
    },
  },
  {
    id: 2,
    title: "Product B",
    handle: "product-b",
    description: "Description for Product B",
    published_at: "2023-01-01",
    created_at: "2023-01-01",
    type: "Type B",
    vendor: "Vendor B",
    tags: ["tag1", "tag3"],
    variants: [],
    images: [],
    options: [],
    shop: "shop.myshopify.com",
    status: "active",
    price: 150,
    price_min: 150,
    price_max: 150,
    available: true,
    compare_at_price: 180,
    inventory_quantity: 5,
    inventory_management: "shopify",
    inventory_policy: "deny",
    fulfillment_service: "manual",
    requires_shipping: true,
    taxable: true,
    weight: 2,
    weight_unit: "kg",
    specs: {
      color: "Blue",
      size: "Large",
      weight: "3kg",
      available_regions: ["US", "EU"],
    },
  },
];

// Mock component that wraps our actual component with the needed props
// This avoids the context issues
const MockMultiColumnComparison = (props: any) => {
  // In a real component, we would use the context values
  // But for Storybook, we'll just render the component with mock props
  return (
    <div style={{ padding: "20px" }}>
      <MultiColumnComparison {...props} />
    </div>
  );
};

const meta = {
  title: "Product Comparison/MultiColumnComparison",
  component: MockMultiColumnComparison,
  parameters: {
    layout: "fullscreen",
  },
  tags: ["autodocs"],
  decorators: [
    (Story) => {
      return (
        <div data-theme="light">
          <Story />
        </div>
      );
    },
  ],
} satisfies Meta<typeof MockMultiColumnComparison>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Empty: Story = {
  args: {
    products: [],
  },
};

export const WithTwoProducts: Story = {
  args: {
    products: sampleProducts,
  },
  decorators: [
    (Story) => {
      // Set the pathname for testing
      // if (typeof window !== 'undefined') {
      //   Object.defineProperty(window, 'location', {
      //     value: { pathname: '/products/product-a' },
      //     writable: true
      //   });
      // }
      return (
        <div data-theme="light">
          <Story />
        </div>
      );
    },
  ],
};

export const WithThreeProducts: Story = {
  args: {
    products: [
      ...sampleProducts,
      {
        id: 3,
        title: "Product C",
        handle: "product-c",
        description: "Description for Product C",
        published_at: "2023-01-01",
        created_at: "2023-01-01",
        type: "Type C",
        vendor: "Vendor C",
        tags: ["tag1", "tag4"],
        variants: [],
        images: [],
        options: [],
        shop: "shop.myshopify.com",
        status: "active",
        price: 80,
        price_min: 80,
        price_max: 80,
        available: true,
        compare_at_price: 100,
        inventory_quantity: 15,
        inventory_management: "shopify",
        inventory_policy: "deny",
        fulfillment_service: "manual",
        requires_shipping: true,
        taxable: true,
        weight: 1.5,
        weight_unit: "kg",
        specs: {
          color: "Green",
          size: "Small",
          weight: "1.5kg",
          available_regions: ["US", "CA", "JP"],
        },
      },
    ],
  },
};
