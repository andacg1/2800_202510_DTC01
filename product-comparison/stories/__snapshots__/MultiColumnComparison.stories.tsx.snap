// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Comparison/MultiColumnComparison Empty smoke-test 1`] = `
<div data-theme="light">
  <div style="padding: 20px;">
    <div>
      <div class="basic-multi-select css-b62m3t-container">
        <span id="react-select-2-live-region"
              class="css-1f43avz-a11yText-A11yText"
        >
        </span>
        <span aria-live="polite"
              aria-atomic="false"
              aria-relevant="additions text"
              role="log"
              class="css-1f43avz-a11yText-A11yText"
        >
        </span>
        <div class="select-internal__control css-1yx4vts-control">
          <div class="select-internal__value-container select-internal__value-container--is-multi css-hlgwow"
               style="display: grid;"
          >
            <div class="select-internal__placeholder css-1jqq78o-placeholder"
                 id="react-select-2-placeholder"
            >
              Select...
            </div>
            <div class="select-internal__input-container css-uji2aa"
                 data-value
            >
              <input class="select-internal__input"
                     autocapitalize="none"
                     autocomplete="off"
                     autocorrect="off"
                     id="react-select-2-input"
                     spellcheck="false"
                     tabindex="0"
                     type="text"
                     aria-autocomplete="list"
                     aria-expanded="false"
                     aria-haspopup="true"
                     role="combobox"
                     aria-activedescendant
                     aria-describedby="react-select-2-placeholder"
                     value
                     style="color: inherit; background: 0px center; opacity: 1; width: 100%; grid-area: 1 / 2; font: inherit; min-width: 2px; border: 0px; margin: 0px; outline: 0px; padding: 0px;"
              >
            </div>
          </div>
          <div class="select-internal__indicators css-1wy0on6">
            <span class="select-internal__indicator-separator css-1u9des2-indicatorSeparator">
            </span>
            <div class="select-internal__indicator select-internal__dropdown-indicator css-1xc3v61-indicatorContainer"
                 aria-hidden="true"
            >
              <svg height="20"
                   width="20"
                   viewbox="0 0 20 20"
                   aria-hidden="true"
                   focusable="false"
                   class="css-tj5bde-Svg"
              >
                <path d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z">
                </path>
              </svg>
            </div>
          </div>
        </div>
        <div>
          <input name="products"
                 type="hidden"
                 value
          >
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Product Comparison/MultiColumnComparison WithThreeProducts smoke-test 1`] = `
<div data-theme="light">
  <div style="padding: 20px;">
    <div>
      <div class="basic-multi-select css-b62m3t-container">
        <span id="react-select-4-live-region"
              class="css-1f43avz-a11yText-A11yText"
        >
        </span>
        <span aria-live="polite"
              aria-atomic="false"
              aria-relevant="additions text"
              role="log"
              class="css-1f43avz-a11yText-A11yText"
        >
        </span>
        <div class="select-internal__control css-1yx4vts-control">
          <div class="select-internal__value-container select-internal__value-container--is-multi css-hlgwow"
               style="display: grid;"
          >
            <div class="select-internal__placeholder css-1jqq78o-placeholder"
                 id="react-select-4-placeholder"
            >
              Select...
            </div>
            <div class="select-internal__input-container css-uji2aa"
                 data-value
            >
              <input class="select-internal__input"
                     autocapitalize="none"
                     autocomplete="off"
                     autocorrect="off"
                     id="react-select-4-input"
                     spellcheck="false"
                     tabindex="0"
                     type="text"
                     aria-autocomplete="list"
                     aria-expanded="false"
                     aria-haspopup="true"
                     role="combobox"
                     aria-activedescendant
                     aria-describedby="react-select-4-placeholder"
                     value
                     style="color: inherit; background: 0px center; opacity: 1; width: 100%; grid-area: 1 / 2; font: inherit; min-width: 2px; border: 0px; margin: 0px; outline: 0px; padding: 0px;"
              >
            </div>
          </div>
          <div class="select-internal__indicators css-1wy0on6">
            <span class="select-internal__indicator-separator css-1u9des2-indicatorSeparator">
            </span>
            <div class="select-internal__indicator select-internal__dropdown-indicator css-1xc3v61-indicatorContainer"
                 aria-hidden="true"
            >
              <svg height="20"
                   width="20"
                   viewbox="0 0 20 20"
                   aria-hidden="true"
                   focusable="false"
                   class="css-tj5bde-Svg"
              >
                <path d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z">
                </path>
              </svg>
            </div>
          </div>
        </div>
        <div>
          <input name="products"
                 type="hidden"
                 value
          >
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Product Comparison/MultiColumnComparison WithTwoProducts smoke-test 1`] = `
<div data-theme="light">
  <div data-theme="light">
    <div style="padding: 20px;">
      <div>
        <div class="basic-multi-select css-b62m3t-container">
          <span id="react-select-3-live-region"
                class="css-1f43avz-a11yText-A11yText"
          >
          </span>
          <span aria-live="polite"
                aria-atomic="false"
                aria-relevant="additions text"
                role="log"
                class="css-1f43avz-a11yText-A11yText"
          >
          </span>
          <div class="select-internal__control css-1yx4vts-control">
            <div class="select-internal__value-container select-internal__value-container--is-multi css-hlgwow"
                 style="display: grid;"
            >
              <div class="select-internal__placeholder css-1jqq78o-placeholder"
                   id="react-select-3-placeholder"
              >
                Select...
              </div>
              <div class="select-internal__input-container css-uji2aa"
                   data-value
              >
                <input class="select-internal__input"
                       autocapitalize="none"
                       autocomplete="off"
                       autocorrect="off"
                       id="react-select-3-input"
                       spellcheck="false"
                       tabindex="0"
                       type="text"
                       aria-autocomplete="list"
                       aria-expanded="false"
                       aria-haspopup="true"
                       role="combobox"
                       aria-activedescendant
                       aria-describedby="react-select-3-placeholder"
                       value
                       style="color: inherit; background: 0px center; opacity: 1; width: 100%; grid-area: 1 / 2; font: inherit; min-width: 2px; border: 0px; margin: 0px; outline: 0px; padding: 0px;"
                >
              </div>
            </div>
            <div class="select-internal__indicators css-1wy0on6">
              <span class="select-internal__indicator-separator css-1u9des2-indicatorSeparator">
              </span>
              <div class="select-internal__indicator select-internal__dropdown-indicator css-1xc3v61-indicatorContainer"
                   aria-hidden="true"
              >
                <svg height="20"
                     width="20"
                     viewbox="0 0 20 20"
                     aria-hidden="true"
                     focusable="false"
                     class="css-tj5bde-Svg"
                >
                  <path d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z">
                  </path>
                </svg>
              </div>
            </div>
          </div>
          <div>
            <input name="products"
                   type="hidden"
                   value
            >
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
